### ISSUE TITLE
Allow sorting in filters

### ISSUE THREAD

**Issue by christiaanwesterbeek** [marmelab/react-admin#2003]
**Is your feature request related to a problem? Please describe.**
In the situation of a [custom iterator](https://marmelab.com/react-admin/List.html#using-a-custom-iterator) in `List` where records are shown as cards. There's no handle to sort because there are no column headers as in a Datagrid.

**Describe the solution you'd like**
I propose a solution where sorting is possible in the [List filters](https://marmelab.com/react-admin/List.html#filters). This would require 2 things (I suppose)
1. For `List` to pass the `setSort` prop to each of the filters [here](https://github.com/marmelab/react-admin/blob/44e9a430356c241ab00a393a6da99f7caf16b3ac/packages/ra-ui-materialui/src/list/List.js#L133-L138).
2. An input in filters would need to take some kind of boolean (noFilter) that tells the ListController to not go and filter onChange when the boolean is set to true.

Then I can build my own alwaysOn input filter inside the filters that would make use of the setSort prop that it would receive from List.

**Describe alternatives you've considered**
In my current solution, I've made a copy of List with just 1 addition, but I don't like to maintain this thing with future versions of List.
![image](https://user-images.githubusercontent.com/465989/42418397-ff0b32fc-829f-11e8-94f7-0d107f0e70a8.png)

**Additional context**
/

**Comment by christiaanwesterbeek** [comment]
In the mean time, I'm not using a custom List. Instead I transform the filter value and make it a sort property.

```
const BoxFilter = props => {
    const parseSortValue = v => {
        if (v === 'new')
            return {
                field: 'accessible',
                order: 'DESC',
            };
        if (v === 'easy')
            return {
                field: 'difficulty_avg',
                order: 'ASC',
            };
        if (v === 'hard')
            return {
                field: 'difficulty_avg',
                order: 'DESC',
            };
    };
    const formatSortValue = v => {
        if (!v) return 'new';
        if (v.field === 'accessible') return 'new';
        if (v.field === 'difficulty_avg')
            return v.order === 'ASC' ? 'easy' : 'hard';
    };

    return (
        <Filter {...props}>
            <MyInputBasedOnRadioButtonGroupInput
                source="__sort"
                parse={parseSortValue}
                format={formatSortValue}
                alwaysOn
            />
            <TextInput
                source="title"
                label="Title"
                alwaysOn
            />
        </Filter>
    );
};
```

And in my ra-postgrest-client I [already](https://github.com/marmelab/react-admin/issues/2000#issuecomment-403000655) ignore filters that start with __ . Now I pick `__sort` from the filters and apply it as sort.
```
    const convertRESTRequestToHTTP = (type, resource, params) => {
        let url = '';
        const options = {};
        options.headers = new Headers();
        switch (type) {
            case GET_LIST: {
                const { page, perPage } = params.pagination;
                const { field, order } =
                    (params.filter && params.filter.__sort) || params.sort;
                //  ^^^^ allow sorting from filters ^^^^^^^
                options.headers.set('Range-Unit', 'items');
                options.headers.set(
                    'Range',
                    `${(page - 1) * perPage}-${page * perPage - 1}`
                );
                options.headers.set('Prefer', 'count=exact');
                const query = {
                    order: `${field}.${order.toLowerCase()}`,
                };
                Object.assign(query, convertFilters(params.filter));
                if (params.columns) {
                    Object.assign(query, {
                        select: Array.isArray(params.columns)
                            ? params.columns.join(',')
                            : params.columns,
                    });
                }
                url = `${apiUrl}/${resource}?${stringify(query)}`;
                break;
            }
            case GET_ONE:
                ...
```

**Comment by fzaninotto** [comment]
Interesting use case, thanks for your suggestion.

Using the List Filters for sorting is a hack IMO. I think it's a good use case for using a custom `ListView` - what you have already done. So we won't support anything more in the core.

**Comment by fzaninotto** [comment]
If I understand Loopback's doc correctly, they use the word 'filter' to describe what everybody else sees as a 'query'. Not the first time IBM tries to reinvent the wheel ;)

**Comment by fzaninotto** [comment]
The first part of your request was implemented in #2028
